sdk/include/opentelemetry/sdk/resource/semantic_conventions.h

```diff
@@ -0,0 +1,136 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "opentelemetry/version.h"
+
+#define OTEL_RESOURCE_GENERATE_SEMCONV_METHOD(attribute_name, attribute_value) \
+  static constexpr const char *GetAttribute##attribute_name() noexcept { return attribute_value; }
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace resource
+{
+/**
+ * Stores the Constants for semantic Attribute names for resources outlined by the OpenTelemetry
+ * specifications. <see
+ * href,"https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md"/>.
+ */
+
+class SemanticConventions final
+{
+
+public:
+  // service attributes
+  OTEL_RESOURCE_GENERATE_SEMCONV_METHOD(ServiceName, "service.name");
```

Comment:
You can do this instead. Take this compile-time hash function:

```cpp
/// <summary>
/// Compile-time constexpr djb2 hash function for strings
/// </summary>
static constexpr uint32_t hashCode(const char *str, uint32_t h = 0)
{
  return (uint32_t)(!str[h] ? 5381 : ((uint32_t)hashCode(str, h + 1) * (uint32_t)33) ^ str[h]);
}

#define CONST_UINT32_T(x) std::integral_constant<uint32_t, (uint32_t)x>::value

#define CONST_HASHCODE(name) CONST_UINT32_T(OPENTELEMETRY_NAMESPACE::utils::hashCode(#name))
```

Then for every attribute, generate a table of values. At compile time - by listing all of them. Sorry for using this [example](https://github.com/microsoft/cpp_client_telemetry/blob/master/lib/tracing/api/module_list) - something like this:
```cpp
// Hash function below is computing module tag from literal at compile-time using C++11 constexpr
#define DBG_MODULE_CTAG(name)       CONST_UINT32_T(djb2_constexpr_hash(#name))
#define DBG_MODULE_TAG(name)        djb2_hash(name)
#define DBG_LOG_NAMETAG(name)       { DBG_MODULE_CTAG(name) , #name } ,
```

So - you created a compile time table that maps your attribute names to unique IDs. You can still refer to original attributes by their aliases. Like `ATTR_ServiceName`, macro-magic permitting.

The goal is that you can do tricks like switch-case, constexpr, and mapping from attribute to unique attribute id. Then instead of many methods, you end up with `attr(name)` or `attr(id)`. That's how Qt5 `tr()` works, and how translations / resources work in Android pretty much. Except that `tr()` requires string literal, and Android resources are all revolving around `getResource(id)`.

In our model, I think, it'd be more appropriate to perform compile-time mapping from `attr("my.attribute.name")` -> returns actual attribute name within given semantic convention. Not many getters. One getter. Lookup table. And preferably, compile-time lookup, not method call.